homeassistant/components/tado/water_heater.py

```diff
@@ -0,0 +1,303 @@
+    SUPPORT_OPERATION_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+    WaterHeaterDevice,
+)
+from homeassistant.const import ATTR_TEMPERATURE, TEMP_CELSIUS
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from . import DOMAIN, SIGNAL_TADO_UPDATE_RECEIVED
+from .const import (
+    CONST_MODE_OFF,
+    CONST_MODE_SMART_SCHEDULE,
+    CONST_OVERLAY_MANUAL,
+    CONST_OVERLAY_TADO_MODE,
+    DATA,
+    TYPE_HOT_WATER,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+MODE_AUTO = "auto"
+MODE_HEAT = "heat"
+MODE_OFF = "off"
+
+WATER_HEATER_MAP_TADO = {
+    "MANUAL": MODE_HEAT,
+    "TIMER": MODE_HEAT,
+    "TADO_MODE": MODE_HEAT,
+    "SMART_SCHEDULE": MODE_AUTO,
+    "OFF": MODE_OFF,
+}
+
+SUPPORT_FLAGS_HEATER = SUPPORT_OPERATION_MODE
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Tado water heater platform."""
+    api_list = hass.data[DOMAIN][DATA]
+    entities = []
+
+    for tado in api_list:
+        for zone in tado.zones:
+            if zone["type"] in [TYPE_HOT_WATER]:
+                entity = create_water_heater_entity(tado, zone["name"], zone["id"])
+                entities.append(entity)
+
+    if entities:
+        add_entities(entities, True)
+
+
+def create_water_heater_entity(tado, name: str, zone_id: int):
+    """Create a Tado water heater device."""
+    capabilities = tado.get_capabilities(zone_id)
+    supports_temperature_control = capabilities["canSetTemperature"]
+
+    if supports_temperature_control and "temperatures" in capabilities:
+        temperatures = capabilities["temperatures"]
+        min_temp = float(temperatures["celsius"]["min"])
+        max_temp = float(temperatures["celsius"]["max"])
+    else:
+        min_temp = None
+        max_temp = None
+
+    entity = TadoWaterHeater(
+        tado, name, zone_id, supports_temperature_control, min_temp, max_temp
+    )
+
+    return entity
+
+
+class TadoWaterHeater(WaterHeaterDevice):
+    """Representation of a Tado water heater."""
+
+    def __init__(
+        self,
+        tado,
+        zone_name,
+        zone_id,
+        supports_temperature_control,
+        min_temp,
+        max_temp,
+    ):
+        """Initialize of Tado water heater entity."""
+        self._tado = tado
+
+        self.zone_name = zone_name
+        self.zone_id = zone_id
+        self._unique_id = f"{zone_id} {tado.device_id}"
+
+        self._device_is_active = False
+        self._is_away = False
+
+        self._supports_temperature_control = supports_temperature_control
+        self._min_temperature = min_temp
+        self._max_temperature = max_temp
+
+        self._target_temp = None
+
+        self._supported_features = SUPPORT_FLAGS_HEATER
+        if self._supports_temperature_control:
+            self._supported_features |= SUPPORT_TARGET_TEMPERATURE
+
+        if tado.fallback:
+            _LOGGER.debug("Default overlay is set to TADO MODE")
```

Comment:
We don't want side effects, eg logging, in init method. Please move it or remove it.

Reply:
I've removed it.